system/jlib/jptree.cpp

```diff
@@ -3176,6 +2818,408 @@ bool PTree::checkPattern(const char *&xxpath) const
+        AttrKeyValue *_cur = (AttrKeyValue *)cur;
+        if (_cur+1 == end)
+            return nullptr;
+        ++_cur;
+        kv = *_cur;
+        return _cur;
+    }
+}
+
+const char *LocalPTree::getAttributeValue(const char *key) const
+{
+    AttrKeyValue *e = (AttrKeyValue *)findAttribute(key);
+    if (e)
+        return e->value;
+    return nullptr;
+}
+
+unsigned LocalPTree::getAttributeCount() const
+{
+    return attributes.ordinality();
+}
+
+bool LocalPTree::removeAttribute(const char *key)
+{
+    AttrKeyValue *e = (AttrKeyValue *)findAttribute(key);
+    if (!e)
+        return false;
+    free((void *)e->key);
+    free((void *)e->value);
+    AttrKeyValue *firstE = &attributes.item(0);
+    unsigned pos = e-firstE;
+    attributes.remove(pos);
+    return true;
+}
+
+void LocalPTree::setAttribute(const char *attr, const char *val)
+{
+    if (!validateXMLTag(attr+1))
+        throw MakeIPTException(-1, "Invalid xml attribute: %s", attr);
+
+    if (!val)
+        val = "";  // cannot have NULL value
+    if (0 == attributes.ordinality())
+    {
+        AttrKeyValue kv;
+        kv.key = strdup(attr);
+        kv.value = strdup(val);
+        attributes.append(kv);
+    }
+    else
+    {
+        AttrKeyValue *e = (AttrKeyValue *)findAttribute(attr);
+        if (e)
+        {
+            if (streq(e->value, val))
+                return;
+            free((void *)e->value);
+            e->value = strdup(val);
+        }
+        else
+        {
+            AttrKeyValue kv;
+            kv.key = strdup(attr);
+            kv.value = strdup(val);
+            attributes.append(kv);
+        }
+    }
+}
+
+///////////////////
+
+unsigned ChildMapAtom::getHashFromElement(const void *e) const
+{
+    CAtomPTree &elem= (CAtomPTree &) (*(IPropertyTree *)e);
+    return elem.queryKey()->queryHash();
+}
+
+unsigned ChildMapAtomNC::getHashFromElement(const void *e) const
+{
+    CAtomPTree &elem= (CAtomPTree &) (*(IPropertyTree *)e);
+    return elem.queryKey()->queryHash();
+}
+
+///////////////////
+
+CAtomPTree::CAtomPTree(const char *_name, byte _flags, IPTArrayValue *_value, ChildMap *_children) : PTree(_flags, _value, _children)
+{
+    CAtomPTree::setName(_name);
+}
+
+CAtomPTree::~CAtomPTree()
+{
+    killAttrs();
+    if (name)
+    {
+        AtomRefTable *kT = isnocase()?keyTableNC:keyTable;
+        kT->releaseKey(name);
+    }
+}
+
+void CAtomPTree::setName(const char *_name)
+{
+    AtomRefTable *kT = isnocase()?keyTableNC:keyTable;
+    HashKeyElement *oname = name;
+    if (!_name)
+        name = nullptr;
+    else
+    {
+        if (!validateXMLTag(_name))
+            throw MakeIPTException(PTreeExcpt_InvalidTagName, ": %s", _name);
+        name = kT->queryCreate(_name);
+    }
+    if (oname)
+        kT->releaseKey(oname);
+}
+
+const char *CAtomPTree::queryName() const
+{
+    return name?name->get():nullptr;
+}
+
+AttrValue *CAtomPTree::newAttrArray(unsigned n)
```

Comment:
Do we significantly gain from this mechanism?  If so should it also be being used by the non-commoned code?